### PR TITLE
Runtime version

### LINKED ISSUE #241: Name/version in runtime
This should add an entry to the API that returns a `Slicable` struct:

```
struct RuntimeVersion {
	spec_name: String,  // "polkadot"
	impl_name: String,  // "parity-polkadot"
	authoring_version: u32,  // "0"
	spec_version: u32,  // "0"
	impl_version: u32,  // "0"
}
```

The fields have very particular meanings:

- `spec_name` identifies the different Substrate runtimes. There'll be at least `polkadot` and `demo`. A different on-chain `spec_name` to that of the native runtime would normally result in node not attempting to sync further.
- `impl_name` Is the name of the implementation of the spec. This is of little consequence for the node and serves only to differentiate code of different implementation teams. For this codebase, it will be `parity-polkadot`. If there were a non-Rust implementation of the Polkadot runtime (e.g. C++), then it would identify itself with an accordingly different `impl_name`.
- `authoring_version` is the version of the authorship interface. _An authoring node will not attempt to author blocks unless this is equal to its native runtime_.
- `spec_version` is the version of the runtime specification. _A full-node will not attempt to use its native runtime in substitute for the on-chain Wasm runtime unless all of `spec_name`, `spec_version` and `authoring_version` are the same between Wasm and native_.
- `impl_version` is the version of the implementation of the specification. Nodes are free to ignore this; it serves only as an indication that the code is different; as long as the other two versions are the same then while the code may be different, it is nonetheless required to do *the same thing*. Non-consensus-breaking optimisations are about the only changes that could be made which would result in only the `impl_version` changing.

NOTE: This should not be thought of as classic Semver (major/minor/tiny). This triplet have different semantics and mis-interpretation could cause problems. In particular: bug fixes should result in an increment of `spec_version` and possibly `authoring_version`, absolutely not `impl_version` since they change the semantics of the runtime.

### PR DESCRIPTION
**Pull request by arkpar**
Closes #241 

Does not have any performance optimisations.
I've also removed `CheckedId` because it is now obsoleted by the version check and should be handled by the substrate, rather than polkadot anyway.

### CODE REVIEW

File: substrate/runtime/version/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+pub trait Trait {
+	const VERSION: RuntimeVersion;
+}
+
+decl_module! {
+	pub struct Module<T: Trait>;
+}
+
+impl<T: Trait> Module<T> {
+	/// Get runtime verions
```
Comment (pepyakin) [inline substrate/runtime/version/src/lib.rs]: typo: `version`?

File: substrate/runtime/version/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+	pub spec_name: MaybeCowString,
+	pub impl_name: MaybeCowString,
+	pub authoring_version: u32,
+	pub spec_version: u32,
+	pub impl_version: u32,
+}
+
+#[cfg(feature = "std")]
+impl RuntimeVersion {
+	/// Check if this version matches other version for calling into runtime.
```
Comment (pepyakin) [inline substrate/runtime/version/src/lib.rs]: can we copy descriptions from the issue? They're pretty descriptive there...

File: substrate/runtime/version/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+// Substrate Demo is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate Demo.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Version module for runtime; Provide a function that returns runtime verion.
```
Comment (pepyakin) [inline substrate/runtime/version/src/lib.rs]: s/verion/version

File: polkadot/runtime/src/lib.rs [hunk 4]
```diff
@@ -102,6 +105,22 @@ pub type Block = generic::Block<Header, UncheckedExtrinsic>;
 #[cfg_attr(feature = "std", derive(Debug, Serialize, Deserialize))]
 pub struct Concrete;
 
+/// Polkadot runtime version.
+pub const VERSION: RuntimeVersion = RuntimeVersion {
+	spec_name: ver_str!("polkatot"),
```
Comment (pepyakin) [inline polkadot/runtime/src/lib.rs]: s/polkatot/polkadot

File: demo/runtime/src/lib.rs [hunk 5]
```diff
@@ -59,6 +62,22 @@ pub use runtime_primitives::BuildStorage;
 /// Concrete runtime type used to parameterize the various modules.
 pub struct Concrete;
 
+/// Polkadot runtime version.
```
Comment (pepyakin) [inline demo/runtime/src/lib.rs]: demo runtime version?

File: substrate/client/src/error.rs [hunk 6]
```diff
@@ -70,6 +70,12 @@ error_chain! {
 			display("Current state of blockchain has invalid authority count value"),
 		}
 
+		/// Invalid state data.
+		VersionInvalid {
```
Comment (gavofyork) [inline substrate/client/src/error.rs]: Should just assume it's the PoC-1 runtime in this case for now, once we ditch the PoC-1 chain that default can be removed.

**Comment by arkpar** [comment]
Addressed all issues

**Comment by gavofyork** [comment]
quoting #241:

> Actually, there is one addendum: authorship should require authoring_version and spec_name to be identical. But not spec_version.

**Comment by rphmeier** [comment]
> I've also removed CheckedId because it is now obsoleted by the version check and should be handled by the substrate, rather than polkadot anyway.

Why? `CheckedId` is a type-safe handle to indicate that a block exists. It says "this block ID corresponds to data we know locally", which is a useful tool and IMO is not obsoleted by the version check, which tells you if a known block corresponds to a certain runtime version.

**Comment by arkpar** [comment]
@rphmeier I don't see why it should be checked by the polkadot API and not the substrate. Also, `CheckedId` does not guarantee anything since the state can be pruned away any time after the check. Does not look safe to me.

**Comment by arkpar** [comment]
@gavofyork Added that last bit

File: substrate/runtime/version/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+
+#[cfg(feature = "std")]
+pub type VersionString = ::std::borrow::Cow<'static, str>;
+#[cfg(not(feature = "std"))]
+pub type VersionString = &'static str;
+
+#[cfg(feature = "std")]
+#[macro_export]
+macro_rules! ver_str {
+	( $y:expr ) => {{ ::std::borrow::Cow::Borrowed($y) }}
```
Comment (gavofyork) [inline substrate/runtime/version/src/lib.rs]: question: why using `Cow` for std?

Comment (arkpar) [reply substrate/runtime/version/src/lib.rs]: So that it can be defined as `const` and deserialised from `Slicable` using the same type.

**Pull request merged**